KMeans clustering 알고리즘 구현 및 Cluster의 hashValue 테스트

donggeonoh opened the pull request:
## 구현내용
### [feat]
- point 에서 가장 가까운 센터를 구하는 메소드 추가
- 센터에 따른 포인트 Classification 메소드 추가
- 분류한 points에서 평균 값을 통한 center를 구하는 메소드 추가
- classification 후 이전과 point가 달라진 점이 없다면 해당 Clusters를 반환
- Cluster hashable 채택
  - Cluster에 대한 hashable 관련 테스트 코드 작성

## 논의사항
- Cluster 의 논리적 오류
  - Cluster의 center에 대하여 classification 된 points 들을 가지고 있는데 points의 평균 값을 통하여 다시 center 값을 정해줘야 함
  - 이 때, classification 된 points는 이전 center 에 대한 것이기 때문에 논리적인 오류 발생

- 특정한 하나의 클러스터의 center 값이 (0, 0) 으로 계산되는 문제
  - 추후 수정 필요

donggeonoh committed:
[feat] Cluster 구조체 및 Coordinate 구조체 hashable 채택

Cluster 구조체는 hash(into:) 로 구현

donggeonoh committed:
[feat] KMeans classification 함수 구현

센터에 따른 클러스터 생성

donggeonoh committed:
[feat] KMeans move 메서드 삭제

move 메서드를 삭제하고 Cluster에서 연산 프로퍼티로 이동

donggeonoh committed:
[feat] KMeans 클러스터를 구하는 알고리즘 구현

좌표와 초기 센터를 가지고 클러스터를 계산하여 반환하는 알고리즘 작성

donggeonoh committed:
Merge branch 'develop' into feature/kmeans-algorithm

eunjeongS2 commented on Project17-C-Map/InteractiveClusteringMap/QuadTree/BoundingBox.swift:
```diff
@@ -43,19 +43,59 @@ struct Coordinate: Equatable {
+    var center: Coordinate
+    var coordinates: [Coordinate]
+    
+    var updateCenter: Coordinate {
+        coordinates.reduce(.zero, +) / Double(coordinates.count)
+    }
```
updateCenter 뭔가 함수이름같네용
센터를 반환하는게 아니라 센터를 바꿔줄 것 같은 이름이에요 !

donggeonoh replied:
center를 computed property로 바꿔주고 updateCenter 프로퍼티는 지웠습니다!

eunjeongS2 commented on Project17-C-Map/InteractiveClusteringMap/VO/KMeansClustering.swift:
```diff
@@ -0,0 +1,90 @@
+        // x와 센터들 중 가장 가까운 거리를 가지는 center의 index를 구함
+        for (idx, center) in centers.enumerated() {
+            let dist = point.distanceTo(center)
+            if dist < nearestDist {
+                minIndex = idx
+                nearestDist = dist
+            }
```
진짜 nit : guard 문으로 할 수 있지 않을까용 ?

donggeonoh replied:
for문 안에서 쓰여서 if로 처리하는게 나아보여요

eunjeongS2 commented on Project17-C-Map/InteractiveClusteringMapTests/KMeansTests.swift:
```diff
@@ -0,0 +1,61 @@
+        XCTAssertNotEqual(cluster1, cluster2)
+    }
+    
+    func test_kmeans_classify() throws {
+        
+    }
```
완성되면 좋을 것 같은데,, ㅜㅜ

donggeonoh replied:
앗 이부분 지울게요 mock data가 필요해서 추후 작성하겠습니다

eunjeongS2 commented on Project17-C-Map/InteractiveClusteringMapTests/MockPOIService.swift:
```diff
@@ -0,0 +1,39 @@
+
+import Foundation
+
+class MockPOIService: POIServicing {
```
테스트할때 사용하시려구 한건가요?? 지금 코드에는 사용한 부분이 없어보이는데,, 잘못본건가용
사용안하면 안올리는게 좋을 것 같아요 ~~

donggeonoh replied:
넵 알겠습니다 ㅠㅠ

rnfxl92 reviewed (approved):
LGTM

eunjeongS2 commented on Project17-C-Map/InteractiveClusteringMap/QuadTree/BoundingBox.swift:
```diff
@@ -43,19 +43,59 @@ struct Coordinate: Equatable {
     
+    func distanceTo(_ other: Coordinate) -> Double {
+        let powX = pow(self.x - other.x, 2.0)
+        let powY = pow(self.y - other.y, 2.0)
+        
+        return sqrt(powX + powY)
```
진짜 nit.. : 위에 center랑 개행이 안맞는거같아용

donggeonoh replied:
수정했습니다!

donggeonoh committed:
[feat] Center의 거리 변화량을 통한 KMeans 알고리즘 구현

이전 Center와 현재 Center의 거리의 총 합 계산
계산한 모든 거리 < 설정한 거리 인 경우 Cluster를 반환하는 KMeans
알고리즘 작성
PR 피드백 반영

donggeonoh committed:
Merge remote-tracking branch 'origin/feature/kmeans-algorithm' into feature/kmeans-algorithm

donggeonoh committed:
location manager 네임 수정 및 Podfile.lock 업데이트

donggeonoh committed:
Merge branch 'develop' into feature/kmeans-algorithm

donggeonoh committed:
test_means_classify 테스트 메소드 삭제

donggeonoh committed:
Merge branch 'feature/kmeans-algorithm' of github.com:boostcamp-2020/Project17-C-Map into feature/kmeans-algorithm

donggeonoh commented:
@eunjeongS2 은정님 리뷰 반영했습니다 확인해주세요!

donggeonoh committed:
[fix] NMapsMap.NMapsGeometry 모듈 제거

eunjeongS2 reviewed (approved):
고생하셨습니다 ~~~~

beggu84 reviewed (commented):
전 이제야 보기 시작했는데, 이상한 점!
1. 프로젝트에 MockPOIService.swift 가 포함되어 있긴 한데, 파일은 없나보네요.
![image](https://user-images.githubusercontent.com/19990608/100326135-eda9c880-300c-11eb-9723-53cf6a971b6b.png)

2. KMeansTests 에서 다음과 같은 에러가 나는군요.
![image](https://user-images.githubusercontent.com/19990608/100326227-0a460080-300d-11eb-83db-55e0500922fc.png)

beggu84 commented on Project17-C-Map/InteractiveClusteringMap/VO/KMeansClustering.swift:
```diff
@@ -0,0 +1,114 @@
+        
+        return clusters
+    }
+    
```
1. 다른 친절한 함수들 처럼 여기(`classify`)도 설명을 써주시면 좋겠네요 : )

2. 그리고, 첫 다섯 라인을 이렇게 축약할 수 있을까요?
```
var clusters = [Cluster](repeating: Cluster(coordinates: []), count: centers.count)
```

3. `classify` 의 의미라서 `classIndex` 라고 명명한 거라면, `classifiedIndex` 혹은, `indexOfNearestCenter` 의 반환값이니 `centerIndex` 라고 하는 것도 좋을 것 같습니다.

donggeonoh replied:
@beggu84
> 프로젝트에 MockPOIService.swift 가 포함되어 있긴 한데, 파일은 없나보네요.

파일이 있긴 한데 프로젝트에 링크?가 안걸린 것 같아요
해당부분 수정하겠습니다!

나머지 리뷰 부분도 참고하여 수정하겠습니다 감사합니다 ㅎㅎ

donggeonoh replied:
@beggu84 
말씀하신 피드백 반영하였습니다!
리뷰 감사합니다 (_ _)

donggeonoh committed:
[test] KMeans 클러스터링 테스트 추가

KMeans의 클러스터링이 잘 동작하는지 Mock 데이터를 넣어 테스트 구현

donggeonoh committed:
[feat] 리뷰 피드백 반영

classify 주석 작성 및 cluster 배열 초기화 코드 수정

donggeonoh committed:
Merge branch 'develop' into feature/kmeans-algorithm

beggu84 reviewed (approved):
늦은 밤까지 고생하셨습니다. (_ _)
